### PR TITLE
Display errors if the bundle cannot be loaded

### PR DESCRIPTION
**Pull request by igmoweb**
This is a bit crazy idea but I created a script that will display an error in case the bundles are not found or cannot be loaded. I believe that falling back into production bundles if the dev ones aren't found is not a good idea as it could confuse developers. This will display an error with some details.

captura de pantalla 2017-11-29 a las 10 25 21

### CODE REVIEW

**Comment by rmccue** [comment]
Looks alright to me.

Where are the styles for this error taken from? Could we copy CRA's error styling instead? (You can use [react-wp-ssr's exception handler](https://github.com/humanmade/react-wp-ssr/blob/a70bb27b4a264155b8edcc9e1d5ca2f0f345dc68/namespace.php#L189) and [CSS](https://github.com/humanmade/react-wp-ssr/blob/master/error-overlay.css) as an easy way to grab that.)

File: loader.php [hunk 1]
```diff
@@ -189,3 +190,110 @@ function enqueue_assets( $directory, $opts = [] ) {
...
+			padding:1rem;
+		}
+	</style>
+	<div class="error-overlay">
+		<div class="wrapper primaryErrorStyle">
+			<div class="overlay">
+				<div class="header">Failed to render</div>
+				<div class="error-content primaryErrorStyle">
+					<p>React bundle could not be loaded</p>
+					<p>Make sure that webpack dev server is running or that you have build your assets</p>
```
Comment (rmccue) [inline loader.php]: `built`

I'd probably change this to something like:
```
Unable to find React asset manifest

react-wp-scripts was unable to find either a development or production asset manifest. 

Run `npm start` to start the development server or `npm run build` to build a production bundle.
```

File: loader.php [hunk 2]
```diff
@@ -189,3 +190,110 @@ function enqueue_assets( $directory, $opts = [] ) {
 		wp_enqueue_style( $opts['handle'] );
 	}
 }
+
+/**
+ * Display an overlay error when the React bundle cannot be loaded. It also stops the execution.
+ *
+ * @param array $details
+ */
+function handle_assets_error( $details = [] ) {
```
Comment (rmccue) [inline loader.php]: Oh, also, we should `trigger_error( ..., E_USER_WARNING )` in production, and not output internal details.

**Comment by igmoweb** [comment]
@rmccue Thanks, should be solved now.

File: loader.php [hunk 3]
```diff
@@ -189,3 +196,116 @@ function enqueue_assets( $directory, $opts = [] ) {
...
+				<div class="header">Failed to render</div>
+				<div class="error-content primaryErrorStyle">
+					Unable to find React asset manifest.
+					<code>react-wp-scripts</code> was unable to find either a development or production asset manifest.
+					Run <code>npm start</code> to start the development server or <code>npm run build</code> to build a production bundle.
+					<ul>
+						<?php foreach ( $details as $label => $value ) : ?>
+							<li><strong><?php echo esc_html( $label ); ?></strong>: <?php echo esc_html( $value ); ?></li>
+						<?php endforeach; ?>
+					</ul>
```
Comment (rmccue) [inline loader.php]: Could we remove the list in favour of plain text? Also, tbh not sure we need all of these entries, `$directory` is probably enough?

File: loader.php [hunk 4]
```diff
@@ -189,3 +196,116 @@ function enqueue_assets( $directory, $opts = [] ) {
...
+				</div>
+			</div>
+		</div>
+	</div>
+	<?php
+	wp_die();
```
Comment (rmccue) [inline loader.php]: This should just be `die()`, otherwise it'll have extra HTML.

File: loader.php (hunk 3 above)
Comment (igmoweb) [reply loader.php]: Honestly, I don't remember why I did that but I removed them completely, the error should be clear enough where is coming from.

File: loader.php (hunk 4 above)
Comment (igmoweb) [reply loader.php]: Done

File: loader.php [hunk 5]
```diff
@@ -144,8 +142,13 @@ function enqueue_assets( $directory, $opts = [] ) {
...
 	if ( empty( $assets ) ) {
-		// TODO: This should be an error condition.
-		return;
+		if ( WP_DEBUG ) {
+			handle_assets_error();
+		}
+		else {
+			// Trigger a warning, but otherwise do nothing.
+			trigger_error( 'React WP Scripts Error: Unable to find React asset manifest', E_USER_WARNING );
+		}
```
Comment (rmccue) [inline loader.php]: Should this also `return`? We could probably always `trigger_error` to avoid the `else` as well.

Comment (igmoweb) [reply loader.php]: Solved!

**Pull request merged**